File: web/src/lib/engine/completion.ts
// SPEC: V32 (≥1 work set done = complete) · V33 (done and asPlanned recorded separately) · Flow 3 (warm-up rows are
// excluded from every count; "—" is a complete set — weight plays no part). Twin: ios/Crew/Engine/Completion.swift.
export interface SetFacts {
  targetReps: number;
  actualReps: number;
  done: boolean;
  isWarmup: boolean;
}

export interface CompletionFacts {
  complete: boolean;
  setsDone: number;
  setsPlanned: number;
  setsAsPlanned: number;
  sets: { done: boolean; asPlanned: boolean }[];
}

export function asPlanned(set: SetFacts): boolean {
  return set.done && set.actualReps >= set.targetReps;
}

export function completionFacts(sets: SetFacts[]): CompletionFacts {
  const work = sets.filter((set) => !set.isWarmup);
  const setsDone = work.filter((set) => set.done).length;
  return {
    complete: setsDone >= 1,
    setsDone,
    setsPlanned: work.length,
    setsAsPlanned: work.filter(asPlanned).length,
    sets: sets.map((set) => ({ done: set.done, asPlanned: asPlanned(set) })),
  };
}

File: ios/Crew/Engine/Completion.swift
// SPEC: V32 (≥1 work set done = complete) · V33 (done and asPlanned separately) · Flow 3 (warm-ups excluded from every
// count; "—" is a complete set). Twin of completion.ts. WRITTEN — UNVERIFIED (needs Mac).

import Foundation

struct SetFacts: Codable, Equatable {
    let targetReps: Int
    let actualReps: Int
    let done: Bool
    let isWarmup: Bool
}

struct SetOutcome: Codable, Equatable {
    let done: Bool
    let asPlanned: Bool
}

struct CompletionFacts: Codable, Equatable {
    let complete: Bool
    let setsDone: Int
    let setsPlanned: Int
    let setsAsPlanned: Int
    let sets: [SetOutcome]
}

enum Completion {
    static func asPlanned(_ set: SetFacts) -> Bool {
        set.done && set.actualReps >= set.targetReps
    }

    static func completionFacts(_ sets: [SetFacts]) -> CompletionFacts {
        let work = sets.filter { !$0.isWarmup }
        let setsDone = work.filter(\.done).count
        return CompletionFacts(
            complete: setsDone >= 1,
            setsDone: setsDone,
            setsPlanned: work.count,
            setsAsPlanned: work.filter(asPlanned).count,
            sets: sets.map { SetOutcome(done: $0.done, asPlanned: asPlanned($0)) }
        )
    }
}
